src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine+Wallet.kt

```diff
@@ -296,24 +298,46 @@ internal fun TradingStateMachine.onChainDelegations(payload: String): StateChang
 }
 
 internal fun TradingStateMachine.onChainUnbonding(payload: String): StateChanges {
-    val response = parser.decodeJsonObject(payload)
-    return try {
-        val unbonding = response?.get("unbondingResponses")?.let {
-            parser.asList(it)
-        } ?: iListOf()
-        this.wallet = walletProcessor.receivedUnbonding(wallet, unbonding)
-        return StateChanges(iListOf(Changes.accountBalances), null)
-    } catch (e: Exception) {
-        StateChanges(iListOf())
+    if (staticTyping) {
+        val response = parser.asTypedObject<OnChainUnbondingResponse>(payload)
+        val oldValue = internalState.wallet.account.unbondingDelegation
+        walletProcessor.processUnbonding(internalState.wallet, response)
+        return if (oldValue != internalState.wallet.account.unbondingDelegation) {
+            StateChanges(iListOf(Changes.accountBalances), null)
+        } else {
+            StateChanges(iListOf())
+        }
+    } else {
+        val response = parser.decodeJsonObject(payload)
+        return try {
+            val unbonding = response?.get("unbondingResponses")?.let {
+                parser.asList(it)
+            } ?: iListOf()
+            this.wallet = walletProcessor.receivedUnbondingDeprecated(wallet, unbonding)
+            return StateChanges(iListOf(Changes.accountBalances), null)
+        } catch (e: Exception) {
+            StateChanges(iListOf())
+        }
     }
 }
 
 internal fun TradingStateMachine.onChainStakingRewards(payload: String): StateChanges {
-    val response = parser.decodeJsonObject(payload)
-    return try {
-        this.wallet = walletProcessor.receivedStakingRewards(wallet, response)
-        return StateChanges(iListOf(Changes.accountBalances), null)
-    } catch (e: Exception) {
-        StateChanges(iListOf())
+    if (staticTyping) {
+        val response = parser.asTypedObject<OnChainStakingRewardsResponse>(payload)
+        val oldValue = internalState.wallet.account.stakingRewards
+        walletProcessor.processStakingRewards(internalState.wallet, response)
```

Comment:
[q] does `oldValue` not get updated when we mutate the internal state via the `process` methods?

Reply:
Yeah, oldValue contains a copy of the old state, and then we can compare and set the "changes set" only if there is a change.